exercises/lamar_moore/ex02_oo_basics/src/Shape.h

```diff
@@ -0,0 +1,34 @@
+#ifndef SHAPE_H
+#define SHAPE_H
+
+namespace Shapes
+{
+	/** List of possible shape types.
+	*/
+	typedef enum
+	{
+		SQUARE = 0,
+		RECTANGLE = 1,
+		CIRCLE = 2,
+		TRIANGLE = 3
+	}ShapeType;
+
+	/** Shape Shape.h "Shape.h"
+		
+		Base interface for all shape types. This includes methods for calculating the perimeter, area
+		and volume of shapes.
+	*/
+	class Shape
+	{
+	public:
+		virtual double calculateArea()=0;
+		virtual double calculatePerimeter()=0;
+		virtual ShapeType getType();
+		virtual int getNumSides();
+	protected:
+		int numSides;
```

Comment:
I have a preference for enforcing What they need to do, which you have done with your pure virtual functions, but now How. 

The protected `numSides` and `ShapeType` implies that derived classes should set these values. But this is not enforced. int for example does not need to be explicitly initialised, so you the derived classes do not have to set `type` or `numSlides` and the base class portion of the end type will behave incorrectly.

Solve this in either of the following ways.

Have a constructor on the base class which takes both int for numSides and the ShapeType for type as arguments. That way there is no default constructor on the base and derived classes have to call the base class constructor and consciously pass the required arguments. At the same time make both member fields const, so that they cannot be overwritten once set.

Alternately, (my preference) make `getType` and `getNumSides` pure virtual and remove the protected fields so that derived classes are forced to implement everything. I prefer this because then the Shape interface is purely and interface, and not a mix of a base class and and interface.